### PR TITLE
New Dataset and OCR variational pipeline

### PR DESCRIPTION
**Pull request by J-Dymond**
This branch includes:

- `src/arc_spice/data/multieurlex_utils.py`
    - Here code for loading and preprocessing the [MultiEURLEX](https://aclanthology.org/2021.emnlp-main.559.pdf) dataset is located.

- `src/arc_spice/variational_pipelines/RTC_variational_pipeline.py`
    - Here the variational pipeline is located. It has `clean_inference` and `variational_inference` functionality. As well as calculating some confidence metrics on the outputs.

- `src/arc_spice/variational_pipelines/dropout_utils.py`
    - This file contains some utility functions for performing MC dropout. 

- `src/arc_spice/eval/classification_error.py` and `src/arc_spice/eval/translation_error.py`
    - These contain some helper functions for calculating errors and uncertainties relating to the two tasks

- `scripts/variational_RTC_example.py`
    - This is a barebones script with example usage of the variational pipeline.

### CODE REVIEW

File: src/arc_spice/pipelines/RTC_pipeline.py
Comment (lannelin) [inline src/arc_spice/pipelines/RTC_pipeline.py:1]: what's the purpose of this file? will we ever use it and not the variational one? should the variational one extend this rather than be completely separate?

**Review by lannelin — CHANGES_REQUESTED** [review]
leaving a partial review before I jump into meetings.

Overall looks good, I've added some comments requesting some small changes.

File: data/MultiEURLEX/data/eurovoc_concepts.json
Comment (lannelin) [inline data/MultiEURLEX/data/eurovoc_concepts.json:1]: I think we'll need to link to license under CC-BY-4.0, maybe add a README to this folder?

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+)
+
+from arc_spice.variational_pipelines.dropout_utils import (
+    count_dropout,
+    dropout_off,
+    dropout_on,
+    set_dropout,
+)
+
+# From huggingface page with model:
```
Comment (lannelin) [inline src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:21]: remove?

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+            # for each metric in a clean inference run (naive_ouputs)
+            for metric in self.naive_outputs[step]:
+                # create a list for each item
+                new_values = [None] * self.n_variational_runs
+                # populate list with items and assign it in new dictionary
+                for run in range(self.n_variational_runs):
+                    new_values[run] = self.var_output[step][run][metric]
+                new_var_dict[step][metric] = new_values
+        # overwrite the existing output dictionary
+        self.var_output = new_var_dict
```
Comment (lannelin) [inline src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:281]: why not return?

Comment (lannelin) [reply src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:281]: same for other fns

**Review by lannelin — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 3]
```diff
@@ -156,3 +157,12 @@ Thumbs.db
 # Common editor files
 *~
 *.swp
+
+
+# project related
+data/Taxi1500*
```
[the remark this answers is not in the document]
Comment (J-Dymond) [reply .gitignore:163]: No you're right, we can remove this

File: src/arc_spice/pipelines/RTC_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            translation = ""
+            text_splits = self.split_inputs(text, ".")
+
+            for split in text_splits:
+                translator_output = self.translator(split)
+                translation += translator_output[0]["translation_text"]
+
+            return translation
+
+        return self.translator(text)[0]["translation_text"]
```
[the remark this answers is not in the document]
Comment (J-Dymond) [reply src/arc_spice/pipelines/RTC_pipeline.py:79]: Yes, I think this assumes a batch size of 1

File: src/arc_spice/pipelines/RTC_pipeline.py
Comment (J-Dymond) [reply src/arc_spice/pipelines/RTC_pipeline.py:1]: As above, I can remove this now, the variational one has the same functionality when used with `clean_inference()`.

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py (hunk 2 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:281]: No particular reason, we could maybe refactor this in a separate issue? Or I can change this before we merge?

File: requirements.txt
[the remark this answers is not in the document]
Comment (lannelin) [reply requirements.txt:1]: i'd remove

File: src/arc_spice/data/multieurlex_utils.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
+import json
+from typing import Union
+
+import torch
+from datasets import load_dataset
+from datasets.formatting.formatting import LazyRow
+from torch.nn.functional import one_hot
+from torch.utils.data import Dataset
+
+with open("data/MultiEURLEX/data/eurovoc_concepts.json", "r") as concepts_file:
```
[the remark this answers is not in the document]
Comment (J-Dymond) [reply src/arc_spice/data/multieurlex_utils.py:10]: I loaded `Dataset` for typehinting, but can remove, `one_hot` is used in the multi_hot encoding function?

File: src/arc_spice/data/multieurlex_utils.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    )
+    # define metadata
+    meta_data = {
+        "n_classes": len(classes),
+        "class_labels": classes,
+        "class_descriptors": descriptors,
+    }
+    # instantiate the preprocessor
+    preprocesser = PreProcesser(lang_pair)
+    # preprocess each split
```
[the remark this answers is not in the document]
Comment (J-Dymond) [reply src/arc_spice/data/multieurlex_utils.py:129]: Fixed this

File: src/arc_spice/data/multieurlex_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (J-Dymond) [reply src/arc_spice/data/multieurlex_utils.py:10]: Made this change

File: src/arc_spice/eval/classification_error.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# hamming accuracy
+# zero-one accuracy
+
+import torch
+
+
+def hamming_accuracy(preds: torch.Tensor, class_labels: torch.Tensor):
```
[the remark this answers is not in the document]
Comment (J-Dymond) [reply src/arc_spice/eval/classification_error.py:7]: Done

File: src/arc_spice/pipelines/RTC_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
+"""
+    Class for the transcription, translation and summarisation pipeline.
+"""
+
+import torch
+from transformers import pipeline
+
+device = (
```
[the remark this answers is not in the document]
Comment (J-Dymond) [reply src/arc_spice/pipelines/RTC_pipeline.py:8]: I've removed this file now

File: src/arc_spice/pipelines/RTC_pipeline.py
Comment (J-Dymond) [reply src/arc_spice/pipelines/RTC_pipeline.py:1]: removed

File: data/MultiEURLEX/data/eurovoc_concepts.json
Comment (J-Dymond) [reply data/MultiEURLEX/data/eurovoc_concepts.json:1]: added `README.md`

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py (hunk 1 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:21]: Yeah, this was left over from old versions of this file. Removed.

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py (hunk 2 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:281]: will do this now

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py [hunk 9]
```diff
@@ -259,26 +256,26 @@ def stack_translator_sentence_metrics(
             ]
         return stacked
 
-    def stack_variational_outputs(self):
+    def stack_variational_outputs(self, var_output):
```
Comment (lannelin) [inline src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:259]: type hints

**Review by lannelin — CHANGES_REQUESTED** [review]
Functionality looks good.

I resolved the merge conflict for .gitignore as it was stopping the CI checks from running. Those are failing at the moment so could you take a look at why @J-Dymond ? I suspect you haven't got pre-commit installed locally when you're committing up. If you haven't, try:
```
pip install -e ".[dev]" # from project dir, installs with dev deps
pre-commit install
pre-commit run --all-files
```

File: src/arc_spice/__init__.py [hunk 10]
```diff
@@ -1,5 +1,7 @@
 """
-ARC-SPICE: Sample Level, Pipeline Introduced Cumulative Errors (SPICE). Investigating methods for generalisable measurement of cumulative errors in multi-model (and multi-modal) ML pipelines.
+ARC-SPICE: Sample Level, Pipeline Introduced Cumulative Errors (SPICE).
+Investigating methods for generalisable measurement of cumulative errors in
+multi-model (and multi-modal) ML pipelines.
 """
 
 from __future__ import annotations
```
Comment (lannelin) [inline src/arc_spice/__init__.py:7]: not sure what this is for?

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py [hunk 11]
```diff
@@ -256,15 +259,15 @@ def stack_translator_sentence_metrics(
             ]
         return stacked
 
-    def stack_variational_outputs(self, var_output):
+    def stack_variational_outputs(self, var_output: dict[str, list[Any]]):
```
Comment (lannelin) [inline src/arc_spice/variational_pipelines/RTC_variational_pipeline.py:262]: add return type

**Review by lannelin — CHANGES_REQUESTED** [review]
nice work! A couple of very minor comments to check but then should be good to merge.

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py (hunk 9 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/RTC_variational_pipeline.py]: These are fixed now

File: requirements.txt
[the remark this answers is not in the document]
Comment (J-Dymond) [reply requirements.txt:1]: Done

File: src/arc_spice/variational_pipelines/RTC_variational_pipeline.py (hunk 2 above)
Comment (J-Dymond) [reply src/arc_spice/variational_pipelines/RTC_variational_pipeline.py]: Done

File: src/arc_spice/__init__.py (hunk 10 above)
Comment (lannelin) [reply src/arc_spice/__init__.py:7]: ah sorry I see you didn't add this line in this PR but I was wondering what L7 was for and think it can be removed. If agree and convenient then please do before merge but happy to approve regardless.
